### PR TITLE
*: fix unexpected error when setting collate for database

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve?

Problem Summary:
This PR tries to fix the two problems related to the charset and collate when creating a schema:
```
tidb> create database db5 collate ascii_bin;
ERROR 1115 (42000): Unknown character set: ''
```
and
```
tidb> create database test1 charset utf8mb4 collate utf8_general_ci;
ERROR 1115 (42000): Unknown character set: 'utf8mb4'
```

### What is changed and how it works?

What's Changed:
Resolve charset and collation in `CreateSchema()` function in ddl_api.go

Some charset/collation functions are also refactored to make the logic clear enough.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

- NA

### Release note 

- Fix some unexpected errors for specifying charset/collate for `CREATE SCHEMA` syntax, fix missing `COLLATE` part for `SHOW CREATE SCHEMA` syntax.

### CODE REVIEW

File: executor/ddl_test.go [hunk 1]
```diff
@@ -333,6 +333,32 @@ func (s *testSuite6) TestCreateDropDatabase(c *C) {
...
+	tk.MustExec("create database charset_test collate utf8_general_ci;")
+	tk.MustQuery("show create database charset_test;").Check(testutil.RowsWithSep("|",
+		"charset_test|CREATE DATABASE `charset_test` /*!40100 DEFAULT CHARACTER SET utf8 COLLATE utf8_general_ci */",
+	))
+	tk.MustExec("drop database charset_test;")
+	tk.MustExec("create database charset_test collate utf8_general_ci;")
+	tk.MustQuery("show create database charset_test;").Check(testutil.RowsWithSep("|",
+		"charset_test|CREATE DATABASE `charset_test` /*!40100 DEFAULT CHARACTER SET utf8 COLLATE utf8_general_ci */",
+	))
+	tk.MustExec("drop database charset_test;")
```
Comment (wjhuang2016) [inline executor/ddl_test.go:356]: The test here is duplicated.

Comment (bb7133) [reply executor/ddl_test.go]: Removed, thanks.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -423,12 +375,28 @@ func setCharsetCollationFlenDecimal(tp *types.FieldType, specifiedCollates []str
 }
 
 // outPriKeyConstraint is the primary key constraint out of column definition. such as: create table t1 (id int , age int, primary key(id));
```
Comment (zimulala) [inline ddl/ddl_api.go:377]: Please update this comment. The format needs to be like `buildColumnAndConstraint  ....`.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1959,6 +1925,33 @@ func isIgnorableSpec(tp ast.AlterTableType) bool {
...
+	chs = def.Tp.Charset
+	coll = def.Tp.Collate
+	if chs != "" && coll == "" {
+		if coll, err = charset.GetDefaultCollation(chs); err != nil {
+			return "", "", errors.Trace(err)
+		}
+	}
+	for _, opt := range def.Options {
+		switch opt.Tp {
+		case ast.ColumnOptionCollate:
```
Comment (zimulala) [inline ddl/ddl_api.go:1939]: Using `if` is enough?

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -1959,6 +1925,33 @@ func isIgnorableSpec(tp ast.AlterTableType) bool {
 	return tp == ast.AlterTableLock || tp == ast.AlterTableAlgorithm
 }
 
+// getCharsetAndCollateInColumnDef will .
```
[the remark this answers is not in the document]
Comment (bb7133) [reply ddl/ddl_api.go:1928]: Thank you, I wrote the code in a hurry, sorry for that.

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -303,36 +312,31 @@ func buildColumnsAndConstraints(ctx sessionctx.Context, colDefs []*ast.ColumnDef
...
 
-// ResolveCharsetCollation will resolve the charset by the order: table charset > database charset > server default charset,
-// and it will also resolve the collate by the order: table collate > database collate > server default collate.
-func ResolveCharsetCollation(tblCharset, tblCollate, dbCharset, dbCollate string) (string, string, error) {
-	if len(tblCharset) != 0 {
-		// tblCollate is not specified by user.
-		if len(tblCollate) == 0 {
-			defCollate, err := charset.GetDefaultCollation(tblCharset)
+// ResolveCharsetCollation will resolve the charset and collate by the order of parameters:
+//   If any given charsetCollate is not empty, the resolved charset and collate will be returned.
```
[the remark this answers is not in the document]
Comment (bb7133) [reply ddl/ddl_api.go:316]: Done

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -303,36 +312,31 @@ func buildColumnsAndConstraints(ctx sessionctx.Context, colDefs []*ast.ColumnDef
...
+//   If all charsetCollates are empty, the default charset and collate will be returned.
```
[the remark this answers is not in the document]
Comment (bb7133) [reply ddl/ddl_api.go:317]: Done

File: ddl/ddl_api.go (hunk 2 above)
Comment (bb7133) [reply ddl/ddl_api.go:377]: OK

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -58,11 +58,20 @@ import (
 
 const expressionIndexPrefix = "_V$"
 
+type charsetCollate struct {
```
Comment (crazycs520) [inline ddl/ddl_api.go:61]: What's different between `charsetCollate ` and `ast.CharsetOpt` struct?

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -422,13 +374,31 @@ func setCharsetCollationFlenDecimal(tp *types.FieldType, specifiedCollates []str
...
+// outPriKeyConstraint is the primary key constraint out of column definition. For example:
+// `create table t1 (id int , age int, primary key(id));`
+func buildColumnAndConstraint(
+	ctx sessionctx.Context,
+	offset int,
+	colDef *ast.ColumnDef,
+	outPriKeyConstraint *ast.Constraint,
+	tblCharset string,
+	tblCollate string,
+) (*table.Column, []*ast.Constraint, error) {
```
Comment (crazycs520) [inline ddl/ddl_api.go:387]: Ignore the charset of the `database`? If the table was created in old TiDB version, the `charset` of the table will be empty?

Comment (crazycs520) [reply ddl/ddl_api.go:387]: Oh, I see it, the charset of the `database` will be resolve in the outside of this function.

File: ddl/ddl_api.go (hunk 7 above)
Comment (bb7133) [reply ddl/ddl_api.go:61]: Good catch! I did try to look for a struct like `ast.CharsetOpt` but failed to get one... so `charsetCollate` is not needed.

File: ddl/ddl_api.go (hunk 8 above)
Comment (bb7133) [reply ddl/ddl_api.go:387]: Yes, the charset of table has been resolved from the database before calling this function.

**Comment by bb7133** [comment]
Failed test case: 
```
FAIL: TestSyncerSimple (0.30s)
syncer_test.go:244: resp key session_ttl_key kvs [key:"session_ttl_key" create_revision:10 mod_revision:10 version:1 value:"session_ttl_val" lease:9154253952256243751 ] length is != 0
```
https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tidb_ghpr_check_2/detail/tidb_ghpr_check_2/31492/pipeline

**Pull request merged**